Improve Charts compatibility with older and upcoming VTK versions.

dcbr opened the pull request:
### Overview

Improve Charts compatibility with older (<9.0) and upcoming (9.2) VTK versions.

<!-- Resolves #2841 -->


### Details

- Support for the upstream VTK Charts changes introduced in [vtk!8618](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8618). Among other things, this allows users to customize the location and dimensions of Pie and Box Charts.
- Chart background falls back to `GetBackgroundBrush` in older VTK versions now (<9.0). This increases compatibility with older versions, although with slightly erroneous coloring (see [vtk#18238](https://gitlab.kitware.com/vtk/vtk/-/issues/18238)).

### Todo

- [x] Fix tests, including #2841.
- [x] ~Support custom background/border coloring of the active chart.~ --> Will do this in #3756.

dcbr commented:
Regarding the tests I have a question: As this PR introduces different behaviour for different VTK versions, should there be separate tests for each version (and how would that look like) or just tests for the latest version/behaviour?

adeak commented:
Thanks! Please also cross-check with #2863 for potential overlaps (I haven't looked at either yet).

dcbr commented:
> Thanks! Please also cross-check with #2863 for potential overlaps (I haven't looked at either yet).

That one looked good (the axis tests part), I'll merge those changes in here once it is accepted.

adeak commented:
> Regarding the tests I have a question: As this PR introduces different behaviour for different VTK versions, should there be separate tests for each version (and how would that look like) or just tests for the latest version/behaviour?

There's existing practice of skipping tests on older versions where functionality is not there, e.g.
https://github.com/pyvista/pyvista/blob/36ab8b59328f08dd76cf8a0f75db0b42555d4461/tests/plotting/test_plotting.py#L48
https://github.com/pyvista/pyvista/blob/36ab8b59328f08dd76cf8a0f75db0b42555d4461/tests/plotting/test_plotting.py#L64

So I think testing for newest is fine. As soon as there's a VTK 9.2.0 release... not sure what to do with the release candidate...

dcbr commented:
Ok, I will proceed like that then 👍
Not sure what the failing "Labeler / triage" test is for? The other documentation and docstringcheck tests fail because they are using vtk<9.2 (they run fine when using the release candidate version).

adeak commented:
I think the labeler task is about https://github.com/pyvista/pyvista/issues/2082, in any case the failure is due to a resource error. I'll just label your PR manually, hopefully the check will become green after rebase and further work.

dcbr commented:
Ready for review. The documentation build should succeed if it uses VTK 9.2. The reason why the coverage is so low is unclear to me at the moment. All of these properties are tested in `test_charts`, but maybe there is a mismatch in the used VTK version here as well?

banesullivan reviewed (approved):
Looks good to me! Thank you for addressing #3224 

FYI, I merged in main to catch this up and re-run CI since there was a failure

banesullivan commented:
@dcbr, the docs are failing from these changes:

```
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartBox:18: WARNING: Exception occurred in plotting pyvista-ChartBox-8
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartBox.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
looking for now-outdated files... none found
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 3969, in size
    raise ValueError(
ValueError: Cannot set ChartBox geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartBox:18: WARNING: Exception occurred in plotting pyvista-ChartBox-11
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartBox.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 3969, in size
    raise ValueError(
ValueError: Cannot set ChartBox geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartPie:18: WARNING: Exception occurred in plotting pyvista-ChartPie-8
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartPie.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 4254, in size
    raise ValueError(
ValueError: Cannot set ChartPie geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartPie:18: WARNING: Exception occurred in plotting pyvista-ChartPie-11
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartPie.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 4254, in size
    raise ValueError(
ValueError: Cannot set ChartPie geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
```

banesullivan commented:
Please note that there are additional failures from #3782

dcbr commented on tests/test_charts.py:
```diff
@@ -23,8 +23,8 @@
 
-# skip all tests if VTK<9.1.0
-if pyvista.vtk_version_info < (9, 1):
+# skip all tests if VTK<9.2.0
+if pyvista.vtk_version_info < (9, 2):
     pytestmark = pytest.mark.skip
```
I didn't touch that except for increasing the minor VTK version, but I can use the decorator style instead.

dcbr replied:
And to answer your question: it does indeed skip the whole module, that's why coverage seems so low (all chart tests skipped as the used VTK version is lower than 9.2). Now I'm actually wondering why this module skip was put there in the first place, some tests for sure should be able to run on older VTK versions (that's the purpose of this PR as well).

dcbr replied:
So the reason this was added was testing problems with some older VTK versions that were missing a certain module (see #2012). I'm going to just keep it there (with the increased version number) to keep things safe and simple, as it worked just fine before and works just fine on the newest VTK version.

dcbr commented:
> @dcbr, the docs are failing from these changes

The docs fail because they are using VTK 9.1.0 (and they need VTK 9.2.0 to showcase all chart features). I believe we need a new osmesa VTK wheel [here](https://github.com/pyvista/pyvista-wheels) and update this line accordingly (probably best to do in a separate PR?):

https://github.com/pyvista/pyvista/blob/55c9e96fb0a0fe6abae783aa91d16ba19c2b59f6/.github/workflows/docs.yml#L49

Additionally, I believe we have all these coverage complaints because also there we are using VTK 9.1.0 instead of the latest version (coupled with `python==3.10`):

https://github.com/pyvista/pyvista/blob/55c9e96fb0a0fe6abae783aa91d16ba19c2b59f6/.github/workflows/testing-and-deployment.yml#L193

akaszynski commented:
> probably best to do in a separate PR?

Yes. We'll need a new wheel. I'm also tempted to run (or try to) the docs in Python 3.11 since it should be somewhat faster.

banesullivan commented:
> Yes. We'll need a new wheel.

I just built and published a wheel for 9.2.5 here: https://github.com/pyvista/pyvista-wheels/blob/main/vtk_osmesa-9.2.5-cp39-cp39-linux_x86_64.whl

```
pip install https://github.com/pyvista/pyvista-wheels/raw/main/vtk_osmesa-9.2.5-cp39-cp39-linux_x86_64.whl
```

This build is from the current (as of writing this comment) `release` branch in VTK which is ~9.2.5. My build configurations make this an easy process: https://github.com/banesullivan/vtk-cmake

```py
git clone git@gitlab.kitware.com:vtk/vtk.git && cd vtk
git checkout release
git submodule update --init --recursive
mkdir build && cd build
# copy the `*.cmake` files from my build configurations
cmake -GNinja -C osmesa.cmake ..
ninja
python setup.py bdist_wheel
```

> I'm also tempted to run (or try to) the docs in Python 3.11 since it should be somewhat faster.

I'll try to upload a 3.11 wheel as well

FYI, VTK is soon to start publishing OSMesa wheels as well now that https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9820 is finished

dcbr commented:
> I'm also tempted to run (or try to) the docs in Python 3.11 since it should be somewhat faster.

I just installed Python 3.11.1 locally (Windows 10) and doc building seems to complete successfully with VTK 9.2.5. I'm getting following warnings during the build (not sure if related to the python or VTK version):

```bash
WARNING: 'pypandoc' not available. Using Sphinx-Gallery to convert rst text blocks to markdown for .ipynb files.
```

<details><summary>Something related to XML reading/writing</summary>

```bash
2023-01-11 22:17:42.886 ( 104.864s) [                ]       vtkXMLWriter.cxx:649    ERR| vtkXMLPolyDataWriter (0000016F6353D180): Error opening output file "C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp"
2023-01-11 22:17:42.893 ( 104.871s) [                ]       vtkXMLWriter.cxx:652    ERR| vtkXMLPolyDataWriter (0000016F6353D180): Error code "Permission denied"
2023-01-11 22:17:42.893 ( 104.871s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC2790): Algorithm vtkXMLPolyDataWriter (0000016F6353D180) returned failure for request: vtkInformation (0000016F638DD080)
  Debug: Off
  Modified Time: 2590781
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1
  FROM_OUTPUT_PORT: -1


2023-01-11 22:17:42.894 ( 104.873s) [                ]       vtkXMLReader.cxx:317    ERR| vtkXMLPolyDataReader (0000016F4B25C860): Error opening file C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp
2023-01-11 22:17:42.895 ( 104.874s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC1FB0): Algorithm vtkXMLPolyDataReader (0000016F4B25C860) returned failure for request: vtkInformation (0000016F638DE660)
  Debug: Off
  Modified Time: 2590874
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1


2023-01-11 22:17:42.899 ( 104.877s) [                ]       vtkXMLReader.cxx:317    ERR| vtkXMLPolyDataReader (0000016F4B25B660): Error opening file C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp
2023-01-11 22:17:42.899 ( 104.877s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC2F70): Algorithm vtkXMLPolyDataReader (0000016F4B25B660) returned failure for request: vtkInformation (0000016F638DD390)
  Debug: Off
  Modified Time: 2591026
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1


2023-01-11 22:17:42.901 ( 104.879s) [                ]       vtkXMLReader.cxx:317    ERR| vtkXMLPolyDataReader (0000016F4B25B660): Error opening file C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp
2023-01-11 22:17:42.901 ( 104.880s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC2F70): Algorithm vtkXMLPolyDataReader (0000016F4B25B660) returned failure for request: vtkInformation (0000016F638DD390)
  Debug: Off
  Modified Time: 2591026
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1


2023-01-11 22:17:42.902 ( 104.881s) [                ]vtkStreamingDemandDrive:819    ERR| vtkCompositeDataPipeline (0000016F63AC2F70): No update piece number has been set in the information for output port 0 on algorithm vtkXMLPolyDataReader (0000016F4B25B660).
```

</details>

```bash
P:\python\venv\pyvista311\Lib\site-packages\zmq\_future.py:679: RuntimeWarning: Proactor event loop does not implement add_reader family of methods required for zmq. Registering an additional selector thread for add_reader support via tornado. Use `asyncio.set_event_loop_policy(WindowsSelectorEventLoopPolicy())` to avoid this warning.
  self._get_loop()
```

```bash
Assertion failed: pfd.revents & POLLIN (C:\projects\libzmq\src\signaler.cpp:265)
```

Finally, this one occurs multiple times throughout the build:
```bash
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
```

Other than that, everything looks good though.

adeak commented:
> WARNING: 'pypandoc' not available. Using Sphinx-Gallery to convert rst text blocks to markdown for .ipynb files.

That one I know (from Python 3.9): it's what it says. Presumably if you install `pypandoc` in your environment this will go away. I never install it (we don't have it in our requirements files), and all this does is raise that warning. I never check the notebooks generated by the doc build, it might differ from what `pypandoc` would do, but anyway this doesn't break the build, you can ignore it.

dcbr commented:
Ready for review. The remaining codecov complaints are all related to "compatibility code" for VTK<9 (which passed before #3822, but fails now that VTK 9.2 is used for coverage reports).

adeak commented:
> The remaining codecov complaints are all related to "compatibility code" for VTK<9 (which passed before #3822, but fails now that VTK 9.2 is used for coverage reports).

Should we add no-cover pragmas? VTK < 9 is becoming pretty obscure.

dcbr commented:
> Should we add no-cover pragmas? VTK < 9 is becoming pretty obscure.

Definitely, I kind of forgot those existed. Added in last commit!
